lib/puppet/ssl/host.rb

```diff
@@ -394,6 +394,37 @@ def state
 
   private
 
+  # Load a previously generated CSR either from memory or from disk
+  # @return [Puppet::SSL::CertificateRequest]
+  def load_certificate_request
+    if Puppet::SSL::CertificateRequest.indirection.terminus_class == :memory
+      return Puppet::SSL::CertificateRequest.indirection.find(cert_name)
+    end
+
+    request_path = certificate_request_location(name)
+    if Puppet::FileSystem.exist?(request_path)
+      Puppet::SSL::CertificateRequest.from_s(Puppet::FileSystem.read(request_path))
+    end
+  end
+
+  # Submit the CSR to the CA, either via an HTTP PUT request, or when testing,
+  # via the indirector. This also caches a copy of the CSR on disk.
+  # @param [Puppet::SSL::CertificateRequest] csr the request to submit
+  def submit_certificate_request(csr)
+    if Puppet::SSL::CertificateRequest.indirection.terminus_class == :memory ||
+      Puppet::SSL::CertificateRequest.indirection.terminus_class == :ca
+      return Puppet::SSL::CertificateRequest.indirection.save(csr)
+    end
+
+    if Puppet::SSL::Host.ca_location == :remote
+      Puppet::Rest::Routes.put_certificate_request(http_client, csr.render, name)
```

Comment:
What are the failure modes here? I don't remember regarding the third party's api of response vs raising. Could we get a response object with a status of 400 if the csr already exists on the master? Or an exception?

I assume throwing an exception if there's trouble here is what would maintain the existing behavior of the CSR.indirection.save, but I'm not positive.

Reply:
Looks like it might be worth adding something to the new client akin to what the indirector used to do with responses, see https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/rest.rb#L282-L314.

Reply:
From functionally testing - removing the Cert and CSR on the agent after having autosigned it on the master - works as expected. It looks like any additional error handling can be done later.